### DISCUSSION TITLE
How do I find the largest white rectangle (bounding box)  in any given image of any size or shape

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#6458] [Digital Image Processing]
Linux Cenos 8 stream
IM7

I have various images of varying sizes. I want the size and coordinates of the largest white area within that image. A sample image is enclosed. The 2nd one shows the rectangle in black

Thanks

![vineborder](https://github.com/ImageMagick/ImageMagick/assets/12046302/d00c6f46-8bc6-446d-a802-47fa04588867)
![vineborder1](https://github.com/ImageMagick/ImageMagick/assets/12046302/62515191-ce0b-4875-8e87-6057d802bf4e)

**Comment by snibgo** [comment]
Windows CMD commands shown.

What is the average of the non-white pixels?
```
D:\temp>%IMG7%magick c:\web\im\ivyBorder.jpg -fuzz 10% -transparent White -scale "1x1^!" -alpha off txt:
# ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (73,138,46)  #498A2E  srgb(28.7065%,54.2131%,17.8847%)
```
Add a border of that colour, and find what the trim would be:
```
D:\temp>%IMG7%magick c:\web\im\ivyBorder.jpg -bordercolor srgb(28.7065%,54.2131%,17.8847%) -border 1 -define trim:percent-background=0 -fuzz 50% -format %@ info:
4214x1614+931+1217
```
Test this by drawing the rectangle:
```
D:\temp>%IMG7%magick c:\web\im\ivyBorder.jpg -fill red -draw "rectangle %[fx:931-1],%[fx:1217-1],%[fx:931+4214-2],%[fx:1217+1614-2]" x1.png
```
[EDIT: added `-2` to right and bottom of rectangle.]

To save bandwidth, here is a small jpg version of x1.png:

![x1](https://github.com/ImageMagick/ImageMagick/assets/33812368/6d72dd3a-ecc8-48de-8f67-fc9ecf121314)

These commands could be combined into one.

Strictly speaking, the result may not be the largest possible rectangle.

**Reply by srini8080** [reply]
As you said, it does not work for certain images like this
![27137](https://github.com/ImageMagick/ImageMagick/assets/12046302/0a75d851-523d-4ff3-a9cf-36ae9aeacf4d)

**Comment by snibgo** [comment]
You can start by making all the roughly non-white pixels black, and the rest exactly white:
```
-fuzz 5% -fill Black +opaque White -fill White +opaque Black
```
... and then continue as I showed above (but see my edit).

Another variation is to use connected-components to find the largest white area. Crop to the bounding box of that area, then continue as I showed above.

My comment about "Strictly speaking, the result may not be the largest possible rectangle" means that, of all possible solutions, the algorithm is not guaranteed to find the largest possible rectangle. It doesn't search the entire space of possible solutions. However, it will find a local maximum, which means that the resulting rectangle can't be extended by one row or column in any direction.

**Reply by srini8080** [reply]
can we do something like turning it into black and white and make a mask and find only the white part, then make the largest rectangle?

**Comment by srini8080** [comment]
convert x1.jpg -transparent black -transparent white   -alpha extract -negate   -colorspace rgb -fill black -opaque black   output.png.
Then I used your first solution on output.png. It worked

**Comment by fmw42** [comment]
Here is one way in Unix syntax.

![img2](https://github.com/ImageMagick/ImageMagick/assets/13458218/463bf9ce-1cff-40f2-b8dd-b4cbb525744a)


```
cropbox=`magick img2.jpg -negate -threshold 0 -negate -background black -define trim:percent-background=0% -format "%@" info:`
echo $cropbox
3856x3136+272+392
magick img2.jpg -crop $cropbox +repage img2_crop.jpg
```

Result below:

![img2_crop](https://github.com/ImageMagick/ImageMagick/assets/13458218/21c08a9c-c581-4287-8ccb-befd95f32572)
